internal/provider/retry.go

```diff
@@ -0,0 +1,202 @@
+}
+
+// WrapRetryOperation converts a RetryOperation interface to a RetryOperationFunc
+func WrapRetryOperation(op RetryOperation) RetryOperationFunc {
+	return func() ([]byte, diag.Diagnostics, int) {
+		return op.Execute()
+	}
+}
+
+// RetryConfig contains configuration for retrying host operations
+type RetryConfig struct {
+	stateConf          *retry.StateChangeConf
+	operationName      string
+	operation          RetryOperationFunc
+	successStatusCodes []int
+	ctx                context.Context
+}
+
+type RetryResult struct {
+	Body  []byte
+	Diags diag.Diagnostics
+	State string
+}
+
+const (
+	// Default Retry Times
+	DefaultRetryTimeout      = 1800 // Overall timeout for retry (seconds) Default: 30min
+	DefaultRetryDelay        = 5    // Time to wait between retries (seconds)
+	DefaultRetryInitialDelay = 2    // Initial delay before first retry (seconds)
+
+	// Retry States
+	RetryStateError    = "error"
+	RetryStateRetrying = "retrying"
+	RetryStateSuccess  = "success"
+)
+
+var (
+	// Success status codes for retry operations
+	DefaultRetrySuccessStatusCodes = []int{http.StatusAccepted, http.StatusNoContent}
+
+	// Retryable status codes for retry operations
+	// Common retryable scenarios based on RFC 7231 and industry standards:
+	// - HTTP 409: Resource conflict (host in use by running jobs)
+	// - HTTP 408: Request timeout
+	// - HTTP 429: Too many requests (rate limiting)
+	// - HTTP 500: Internal server error
+	// - HTTP 502: Bad gateway
+	// - HTTP 503: Service unavailable
+	// - HTTP 504: Gateway timeout
+	DefaultRetryableStatusCodes = []int{http.StatusConflict, http.StatusRequestTimeout,
+		http.StatusTooManyRequests, http.StatusInternalServerError, http.StatusBadGateway,
+		http.StatusServiceUnavailable, http.StatusGatewayTimeout}
+)
+
+// SafeDurationFromSeconds safely converts seconds to time.Duration, checking for overflow
+func SafeDurationFromSeconds(seconds int64) (time.Duration, error) {
+	// Maximum duration in seconds for int64 is roughly 292 years
+	const maxDurationSeconds = math.MaxInt64 / int64(time.Second)
+	if seconds < 0 {
+		return 0, fmt.Errorf("duration must be non-negative, got: %d seconds", seconds)
+	}
+	if seconds > maxDurationSeconds {
+		return 0, fmt.Errorf("duration overflow: %d seconds exceeds maximum allowed duration", seconds)
+	}
+
+	return time.Duration(seconds) * time.Second, nil
+}
+
+// CreateRetryConfig creates a RetryConfig wrapping Terraform's retry.StateChangeConf object
+func CreateRetryConfig(ctx context.Context, operationName string, operation RetryOperationFunc,
+	successStatusCodes []int, retryableStatusCodes []int, retryTimeout int64, initialDelay int64,
+	retryDelay int64) (*RetryConfig, diag.Diagnostics) {
+	var diags diag.Diagnostics
+
+	if operation == nil {
+		diags.AddError(
+			"Error configuring retry",
+			"Retry function is not defined",
+		)
+		return nil, diags
+	}
+
+	if len(successStatusCodes) == 0 || successStatusCodes == nil {
```

Comment:
Checking `len(mySlice) == 0` should cover the `nil` case too. Can this be simplified to the following?

```suggestion
	if len(successStatusCodes) == 0 {
```

Same for the check on `retryableStatusCodes` below.